Assets/MixedRealityToolkit.Services/InputAnimation/InputAnimation.cs

```diff
@@ -0,0 +1,710 @@
+                    yield return curves.PositionX;
+                    yield return curves.PositionY;
+                    yield return curves.PositionZ;
+                    yield return curves.RotationX;
+                    yield return curves.RotationY;
+                    yield return curves.RotationZ;
+                    yield return curves.RotationW;
+                }
+                yield return animation.cameraCurves.PositionX;
+                yield return animation.cameraCurves.PositionY;
+                yield return animation.cameraCurves.PositionZ;
+                yield return animation.cameraCurves.RotationX;
+                yield return animation.cameraCurves.RotationY;
+                yield return animation.cameraCurves.RotationZ;
+                yield return animation.cameraCurves.RotationW;
+            }
+
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return this.GetEnumerator();
+            }
+        }
+
+        private IEnumerable<AnimationCurve> AnimationCurves => new AnimationCurveEnumerable(this);
+
+        /// <summary>
+        /// Find an index i in the sorted events list, such that events[i].time <= time < events[i+1].time.
+        /// </summary>
+        /// <returns>
+        /// 0 <= i < eventCount if a full interval could be found.
+        /// -1 if time is less than the first event time.
+        /// eventCount-1 if time is greater than the last event time.
+        /// </returns>
+        /// <remarks>
+        /// Uses binary search.
+        /// </remarks>
+        private static int FindKeyframeInterval(AnimationCurve curve, float time)
+        {
+            var keys = curve.keys;
+            int lowIdx = -1;
+            int highIdx = keys.Length;
+            while (lowIdx < highIdx - 1)
+            {
+                int midIdx = (lowIdx + highIdx) >> 1;
+                if (time >= keys[midIdx].time)
+                {
+                    lowIdx = midIdx;
+                }
+                else
+                {
+                    highIdx = midIdx;
+                }
+            }
+            return lowIdx;
+        }
+
+        /// <summary>
+        /// Add a user-defined marker.
+        /// </summary>
+        public void AddMarker(InputAnimationMarker marker)
+        {
+            int index = FindMarkerInterval(marker.time) + 1;
+            markers.Insert(index, marker);
+        }
+
+        /// <summary>
+        /// Remove the user-defined marker at the given index.
+        /// </summary>
+        public void RemoveMarker(int index)
+        {
+            markers.RemoveAt(index);
+        }
+
+        /// <summary>
+        /// Change the time of the marker at the given index.
+        /// </summary>
+        public void SetMarkerTime(int index, float time)
+        {
+            InputAnimationMarker marker = markers[index];
+            markers.RemoveAt(index);
+
+            int newIndex = FindMarkerInterval(time) + 1;
```

Comment:
This function is basically just a Remove and then Add.
In order to reduce code duplication, consider just doing:

InputAnimationMarker marker = markers[index];
RemoveMarker(index);
AddMarker(marker);

Reply:
Plus that way if you ever get an off by one error (especially dealing with that new index + 1 business) assuming Add/Remove works, then SetMarkerTime will always work.

Reply:
Not quite, the marker can get inserted in the middle of the list, so that the time values remain sorted. If i just use markers.Add(marker) it can jumble the time values.